paasta_tools/kubernetes_tools.py

```diff
@@ -1496,17 +1497,77 @@ def get_readiness_probe(
         else:
             return self.get_liveness_probe(service_namespace_config)
 
-    def get_kubernetes_container_termination_action(self) -> V1LifecycleHandler:
-        command = self.config_dict.get("lifecycle", KubeLifecycleDict({})).get(
-            "pre_stop_command", []
+    def get_lifecycle_dict(self) -> KubeLifecycleDict:
+        return self.config_dict.get("lifecycle", KubeLifecycleDict({}))
+
+    def get_prestop_sleep_seconds(self, is_in_smartstack: bool) -> int:
+        if is_in_smartstack:
+            default = 30
+        else:
+            default = 0
+        return self.get_lifecycle_dict().get("pre_stop_drain_seconds", default)
+
+    def get_hacheck_prestop_sleep_seconds(self) -> int:
+        """The number of seconds to sleep between hadown and terminating the hacheck container. We want hacheck to be
+        up for slightly longer than the main container is, so we default to pre_stop_drain_seconds + 1.
+
+        It doesn't super matter if hacheck goes down before the main container -- if it's down, healthchecks will fail
+        and the service will be removed from smartstack, which is the same effect we get after running hadown.
+        """
+
+        # Everywhere this value is currently used (hacheck sidecar or gunicorn sidecar), we can pretty safely
+        # assume that the service is in smartstack.
+        return self.get_prestop_sleep_seconds(is_in_smartstack=True) + 1
```

Comment:
honestly should probably just go back to the constant 31s. I'm pretty sure it doesn't matter if the hacheck container goes down -- we want the healthchecks to be failing anyway.